apps/api/src/services/agent-manager.service.ts

```diff
@@ -1030,6 +1069,123 @@ export class AgentManager {
     };
   }
 
+  /**
+   * Attach agent-specific metrics to a competition object
+   * @param competition The competition object to enhance
+   * @param agentId The agent ID for metrics calculation
+   * @returns Enhanced competition with agent metrics
+   */
+  async attachCompetitionMetrics(
+    competition: SelectCompetition,
+    agentId: string,
+  ): Promise<EnhancedCompetition> {
+    try {
+      // Get portfolio value (latest snapshot for agent in competition)
+      const agentSnapshots = await getAgentPortfolioSnapshots(
+        competition.id,
+        agentId,
+      );
+      const latestSnapshot = agentSnapshots?.[0]; // Already ordered by timestamp desc
+      const portfolioValue = latestSnapshot
+        ? Number(latestSnapshot.totalValue)
+        : 0;
+
+      // Calculate PnL (similar to calculateAgentMetrics pattern)
+      let pnl = 0;
+      let pnlPercent = 0;
+      if (agentSnapshots && agentSnapshots.length > 1) {
+        const startingValue = Number(
+          agentSnapshots[agentSnapshots.length - 1]?.totalValue || 0,
+        );
+        const currentValue = Number(agentSnapshots[0]?.totalValue || 0);
+        pnl = currentValue - startingValue;
+        pnlPercent = startingValue > 0 ? (pnl / startingValue) * 100 : 0;
+      }
+
+      // Get total trades for agent in this competition
+      const totalTrades = await countAgentTradesInCompetition(
+        agentId,
+        competition.id,
+      );
+
+      // Get agent's ranking in this competition
+      const bestPlacement = await getAgentCompetitionRanking(
+        agentId,
+        competition.id,
+      );
+
+      return {
+        ...competition,
+        portfolioValue,
+        pnl,
+        pnlPercent,
+        totalTrades,
+        bestPlacement,
+      };
+    } catch (error) {
+      console.error(
+        `[AgentManager] Error attaching competition metrics for agent ${agentId} in competition ${competition.id}:`,
+        error,
+      );
+      // Return competition with default values on error
+      return {
+        ...competition,
+        portfolioValue: 0,
+        pnl: 0,
+        pnlPercent: 0,
+        totalTrades: 0,
+        bestPlacement: { rank: 0, totalAgents: 0 },
+      };
+    }
+  }
+
+  /**
+   * Sort competitions by computed fields
+   * @param competitions Array of competitions with metrics
+   * @param sortString Sort string (e.g., "portfolioValue", "-pnl")
+   * @returns Sorted competitions array
+   */
+  private sortCompetitionsByComputedField(
```

Comment:
note to self: revisit this logic in https://github.com/recallnet/js-recall/issues/620 

(what you have is good / necessary. we just need to figure out a solution for "load _all_ results in the db query because the post-processing needs to sort the fields, too")